File: CSX/Closure, Scope & Execution Context/defineFirstArg.js
// DEFINEFIRSTARG

// expected input: callbackFn, arg
// expected output: function

// approach:

// function defineFirstArg(callbackFn, arg)
//  nested function(arg, ...args)
//    return callbackFn(arg, ...args)
//  return nested function

function defineFirstArg(cb, arg) {
  function definedArg(...args) {
    return cb(arg, ...args);
  }
  return definedArg;
}

const subtract = function(big, small) { return big - small; };
const subFrom20 = defineFirstArg(subtract, 20);
console.log(subFrom20(5)); // should log: 15

File: CSX - Python/Closure, Scope & Execution Context/define_first_arg.py
def define_first_arg(cb, arg1):
    def inner(*args):
        return cb(arg1, *args)
    return inner

def subtract(big, small):
    return big - small

sub_from_20 = define_first_arg(subtract, 20)

print(sub_from_20(5))
